bauble/src/value.rs

```diff
@@ -1082,17 +1109,111 @@ impl<'a> Symbols<'a> {
+            }
+        });
+
+        if delayed.is_empty() {
+            return Ok(());
+        }
+
+        // If the length didn't change we have errors.
+        if delayed.len() == old_len {
+            let mut graph = petgraph::graphmap::DiGraphMap::new();
+            let mut map = HashMap::new();
+            for a in delayed.iter() {
+                let node_a = graph.add_node(a.asset.as_ref().map(|p| p.borrow()));
+                map.insert(node_a, a.reference.as_ref());
+
+                for b in delayed.iter() {
+                    if a.reference.could_be(b.asset.borrow()) {
+                        graph.add_edge(node_a, b.asset.as_ref().map(|p| p.borrow()), ());
+                    }
+                }
+            }
+
+            let mut errors = Vec::new();
+            for scc in petgraph::algo::tarjan_scc(&graph) {
+                if scc.len() == 1 {
+                    if map[&scc[0]].could_be(scc[0].borrow()) {
+                        errors.push(
+                            ConversionError::Cycle(vec![(
+                                scc[0].to_string().spanned(scc[0].span),
+                                vec![map[&scc[0]].to_string().spanned(map[&scc[0]].span)],
+                            )])
+                            .spanned(scc[0].span),
+                        )
+                    } else {
+                        errors.push(
+                            ConversionError::RefError(Box::new(RefError {
+                                // TODO: Could pass uses here for better suggestions.
+                                uses: None,
+                                path: map[&scc[0]].value.clone(),
+                                path_ref: PathReference::empty(),
+                                kind: RefKind::Asset,
+                            }))
+                            .spanned(map[&scc[0]].span),
+                        )
+                    }
+                } else {
+                    errors.push(
+                        ConversionError::Cycle(
+                            scc.iter()
+                                .map(|s| {
+                                    (
+                                        s.to_string().spanned(s.span),
+                                        vec![map[s].to_string().spanned(map[s].span)],
+                                    )
+                                })
+                                .collect(),
+                        )
+                        .spanned(scc[0].span),
+                    );
+                }
+            }
+
+            return Err(errors);
+        }
+    }
+}
+
 pub fn register_assets(
     path: TypePath<&str>,
     ctx: &mut crate::context::BaubleContext,
     default_uses: impl IntoIterator<Item = (TypePathElem, PathReference)>,
     values: &Values,
-) -> std::result::Result<(), Vec<Spanned<ConversionError>>> {
+) -> std::result::Result<Vec<DelayedRegister>, Vec<Spanned<ConversionError>>> {
```

Comment:
I see the delayed things are ran immediately after this function is invoked, which makes me question why this function doesn't just invoke `resolve_delayed` itself? Is it also possible that `resolve_delayed` may cause new delayed refs so it has to be invoked in a loop. because it seems to be a delayed ref is just what happens when a reference is pointing towards a value which hasn't yet been resolved, but what if the reference points to an unresolved reference? Is this possible?

Reply:
We call `register_assets` in a loop with all files. Then call `resolve_delayed` with the delayed assets from all those files. And no we don't need to call `resolve_delayed` in a loop, it handles cycles by itself.